### PR TITLE
feat(grouping): Store grouping-method-specific metadata in `GroupHashMetadata`

### PR DESCRIPTION
**Pull request by lobsterkatie**
This is a follow-up to https://github.com/getsentry/sentry/pull/80531 (which added a new `hashing_metaadata` field to the `GroupHashMetadata` table), adding code to actual store data in the new field.

### CODE REVIEW

File: src/sentry/grouping/ingest/grouphash_metadata.py [hunk 1]
```diff
@@ -71,8 +103,16 @@ def create_or_update_grouphash_metadata(
...
-def _get_hash_basis(event: Event, project: Project, variants: dict[str, BaseVariant]) -> HashBasis:
-    main_variant = (
+def get_hash_basis_and_metadata(
+    event: Event, project: Project, variants: dict[str, BaseVariant]
+) -> tuple[HashBasis, HashingMetadata]:
+    hashing_metadata: HashingMetadata = {}
+
+    # TODO: This (and `contributing_variant` below) are typed as `Any` so that we don't have to cast
+    # them to whatever specific subtypes of `BaseVariant` and `GroupingComponent` (respectively)
+    # each of the helper calls below requires. Casting once, to a type retrieved from a look-up,
```
Comment (JoshFerge) [inline src/sentry/grouping/ingest/grouphash_metadata.py:113]: I believe the way to do this as written is to type narrow within the various if functions. e.g. on the first condition:

```
    if hash_basis == HashBasis.STACKTRACE:
        if not isinstance(
            contributing_component,
            StacktraceGroupingComponent
            | ExceptionGroupingComponent
            | ChainedExceptionGroupingComponent
            | ThreadsGroupingComponent,
        ):
            raise ValueError(
                "Contributing component is not a StacktraceGroupingComponent for stacktrace basis"
            )
        if not isinstance(contributing_variant, ComponentVariant):
            raise ValueError("Contributing variant is not a ComponentVariant for stacktrace basis")

        hashing_metadata = _get_stacktrace_hashing_metadata(
            contributing_variant, contributing_component
        )
```

I would argue for replacing [conditional with polymorphism](https://refactoring.guru/replace-conditional-with-polymorphism) as an alternative, but understand that given existing grouping code, this way is easier.

**Review by JoshFerge — APPROVED** [review]
the PR looks good. included one idea to fix some of the Any typing. Could in the future look to replace more of the conditional logic with methods on the various `Variant` and `Component` classes. I have not reviewed the snapshots, and will do so tomorrow to spot check outputs. it feels kind of weird that there's no regular tests of this, but i assume in grouping code we let snapshots cover things generally.

File: src/sentry/grouping/ingest/grouphash_metadata.py (hunk 1 above)
Comment (lobsterkatie) [reply src/sentry/grouping/ingest/grouphash_metadata.py:113]: Oh, yeah... something like this, probably?

```python
class BaseHashBasis[
    VariantType: BaseVariant,
    ComponentType: BaseGroupingComponent[Any] | None,
    MetadataType: HashingMetadata,
]:
    name: str
    variants: dict[str, BaseVariant]
    contributing_variant: VariantType
    contributing_component: ComponentType | None

    def __init__(
        self,
        variants: dict[str, BaseVariant],
        contributing_variant: VariantType,
        contributing_component: ComponentType | None = None,
    ):
        self.variants = variants
        self.contributing_variant = contributing_variant
        self.contributing_component = contributing_component

        if contributing_component:
            self.subcomponents_by_id = {
                subcomponent.id: subcomponent for subcomponent in contributing_component.values
            }

    def get_hashing_metadata(self) -> HashingMetadata:
        raise NotImplementedError()


class TemplateHashBasis(
    BaseHashBasis[ComponentVariant, TemplateGroupingComponent, TemplateHashingMetadata]
):
    name: str = "template"

    def get_hashing_metadata(self):
        metadata: TemplateHashingMetadata = {}

        if self.subcomponents_by_id["filename"].values:
            metadata["template_name"] = self.subcomponents_by_id["filename"].values[0]
        if self.subcomponents_by_id["context-line"].values:
            metadata["template_context_line"] = self.subcomponents_by_id["context-line"].values[0]

        return metadata

def get_hash_basis_and_metadata(
    event: Event, project: Project, variants: dict[str, BaseVariant]
) -> tuple[HashBasis, HashingMetadata]:
    # Skipping some of the details below, but you get the idea
    contributing_variant = ...
    contributing_component = ...

    hash_basis, hash_basis_type = GROUPING_METHODS_BY_DESCRIPTION[method_description]

    hashing_metadata = hash_basis_type(
        variants, contributing_variant, contributing_component
    ).get_hashing_metadata()

    return hash_basis, hashing_metadata 
```

Would the typechecker be any happier with that, though? Specifically, wouldn't I still have to type `contributing_variant` and `contributing_component` as `Any` in order to pass them to the `TemplateHashBasis`/etc constructor?

Comment (JoshFerge) [reply src/sentry/grouping/ingest/grouphash_metadata.py:113]: I was thinking moreso putting formatting methods on the variant / component classes, and aiming to remove hash-basis as its own enum if possible. i don't have time right now to go into details, so please feel free to ignore, or look at the type narrowing using `isinstance`. whatever's easiest and can help get this merged faster.

**Comment by lobsterkatie** [comment]
> replace more of the conditional logic with methods on the various `Variant` and `Component` classes

The problem with putting it there is there's not a nice 1-on-1 correspondence between hash basis and either of those class families. This is from one of my attempts along the way:

```python
VARIANT_AND_COMPONENT_TYPES_BY_HASH_BASIS = {
    HashBasis.STACKTRACE: (
        ComponentVariant,
        StacktraceGroupingComponent
        | ExceptionGroupingComponent
        | ChainedExceptionGroupingComponent
        | ThreadsGroupingComponent,
    ),
    HashBasis.MESSAGE: (
        ComponentVariant,
        MessageGroupingComponent | ExceptionGroupingComponent | ChainedExceptionGroupingComponent,
    ),
    HashBasis.FINGERPRINT: (
        CustomFingerprintVariant | SaltedComponentVariant,
        None,
    ),
    HashBasis.SECURITY_VIOLATION: (
        ComponentVariant,
        CSPGroupingComponent
        | ExpectCTGroupingComponent
        | ExpectStapleGroupingComponent
        | HPKPGroupingComponent,
    ),
    HashBasis.TEMPLATE: (ComponentVariant, TemplateGroupingComponent),
    HashBasis.CHECKSUM: (ChecksumVariant, None),
    HashBasis.FALLBACK: (BaseVariant, None),
}
```

You'd end up still needing conditionals, just in the various classes methods, wouldn't you? Like, `ComponentVariant` would need to check to see which of stacktrace, message, violation, and template it was dealing with, and then we're back to where we started, no?

**Comment by JoshFerge** [comment]
> You'd end up still needing conditionals, just in the various classes' methods, wouldn't you? Like, `ComponentVariant` would need to check to see which of stacktrace, message, violation, and template it was dealing with, and then we're back to where we started, no?

I believe the trick would be to set up HashBasis as an actual class, and set it on the variant during initialization, then you can call `Variant.get_metadata` which then references the particular hash basis method here. i understand there is some complcation / nuance around this, and as designed the current implementation doesn't quite allow for this nicely.

(You still end up having conditional logic, but the idea is its all encapsulated in a builder / factory, and in theory, simplifies, as after the objects have been created, you can reference methods safely)

No need to worry about this for now!

**Pull request merged**